internal/mempool/mempool_test.go

```diff
@@ -3172,3 +3172,65 @@ func TestMiningView(t *testing.T) {
 			"even though it was removed from the mempool.")
 	}
 }
+
+// TestStagedTransactionHeight verifies that the height of a transaction
+// that moves from the stage pool into the main pool is set to the height it
+// was initially added to the mempool, rather than the height it was unstaged.
+func TestStagedTransactionHeight(t *testing.T) {
+	harness, spendableOuts, err := newPoolHarness(chaincfg.MainNetParams())
+	if err != nil {
+		t.Fatalf("unable to create test pool: %v", err)
+	}
+
+	txA, _ := harness.CreateSignedTx([]spendableOutput{
+		spendableOuts[0],
+	}, 1)
+
+	ticket, err := harness.CreateTicketPurchase(txA, 40000)
+	if err != nil {
+		t.Fatalf("unable to create ticket purchase transaction %v", err)
+	}
+
+	allTxns := []*dcrutil.Tx{txA, ticket}
+	for index, tx := range allTxns {
+		_, err := harness.txPool.ProcessTransaction(tx,
+			true, false, true, 0)
+		if err != nil {
+			t.Fatalf("ProcessTransaction: failed to accept valid "+
+				"transaction at index %d: %v", index, err)
+		}
+	}
+
+	initialBlockHeight := harness.chain.BestHeight()
+	poolTxDescs := harness.txPool.TxDescs()
+	if len(poolTxDescs) != 1 {
+		t.Fatalf("expected to find exactly one transaction in the mempool but "+
+			"got %v", len(poolTxDescs))
+	}
+
+	poolTxA := poolTxDescs[0]
+	if poolTxA.Height != initialBlockHeight {
+		t.Fatalf("expected txA mempool height to be %v but got %v",
+			initialBlockHeight, poolTxA.Height)
+	}
+
+	// Remove txA, which should bring the ticket out of the stage pool and
+	// into the main pool.
+	newBlockHeight := initialBlockHeight + 1
+	harness.AddFakeUTXO(txA, newBlockHeight)
+	harness.chain.SetHeight(newBlockHeight)
+	harness.txPool.RemoveTransaction(txA, false, noTreasury)
+	harness.txPool.MaybeAcceptDependents(txA, noTreasury)
+
+	poolTxDescs = harness.txPool.TxDescs()
+	if len(poolTxDescs) != 1 {
+		t.Fatalf("expected to find exactly one transaction in the mempool but "+
+			"got %v", len(poolTxDescs))
+	}
+
+	poolTicket := poolTxDescs[0]
+	if poolTicket.Height != initialBlockHeight {
```

Comment:
This probably should make sure that `poolTicket.Tx` is actually the `ticket` transaction too.

Reply:
Updated.